### ISSUE TITLE
Add scorer utility for aligner and for fastspeech2 models. (Feature Request)

### ISSUE THREAD

**Issue by michael-conrad** [DigitalPhonetics/IMS-Toucan#17]
It would be useful to have a scorer utility to process all the training data after training and output the score, audio path, and transcript text to a file.

The purpose would be to identify problematic training data that may need transcript correction or removal due to bad audio or other issues.

This would be useful to remove bad data from future training cycles to achieve better models.

Separate utilities for alignment and the fastspeech models would be ideal.

**Comment by Flux9665** [comment]
I think the utility you want is already there, but a little hidden. 

When you create a fastspeech dataset, then there's the parameter `ctc_selection`. If this is True, then the ctc loss for each sample in the dataset will be collected and all samples with a ctc loss that is one standard deviation higher than the average will be excluded. For small datasets, this should be off, but for big datasets, this usually removes the few thousand worst aligned samples, which usually catches some flawed labels. https://github.com/DigitalPhonetics/IMS-Toucan/blob/81075a6d3a1a048f6f02ccda87579c896337a4fb/TrainingInterfaces/Text_to_Spectrogram/FastSpeech2/FastSpeechDatasetLanguageID.py#L147

And in the meta checkpoint training pipeline there's a function `find_and_remove_faulty_samples` which takes a model, path to weights for the model and a list of datasets as the input. It will then go though all of thesamples in each of the datasets, take note of the loss and if there are samples which have NaN as their loss, it will remember their index and then remove them from the dataset later and overwrite the dataset cache with the cleaned version. I put this function in the meta training pipeline because that was the only place I ever needed it, but it's not at all specific to this model, I should put it in a more general place, like Utilits/utils.py. Will do that in the next version. https://github.com/DigitalPhonetics/IMS-Toucan/blob/81075a6d3a1a048f6f02ccda87579c896337a4fb/TrainingInterfaces/TrainingPipelines/FastSpeech2_MetaCheckpoint.py#L193

So both of those utilities automatically remove samples that might be problematic. But if you have few datapoints, then you probably want to inspect those rather than simply remove them and forget about them. It will tell you about the index of those samples in the datapoint, but since the index is not directly liked to a filename, it might be difficult to inspect them, especially since they will be removed from the dataset afterwards anyways. Having some utility to give you a summary of filenames with NaN losses and filenames with unusually high losses makes sense. I will make something like this, probably in the next version. Maybe until then you can use the two functions I linked to build something that fits your needs.

**Comment by michael-conrad** [comment]
Yeah.. definitely in need of that utility to use to scan through and manually cleanup the data at some point in the future.

I trained to step 2,000 and the loss went NaN.

So backtracked, and enabled the remove bad data points, using the step 1,000 as the checkpoint.

Had 5 bad data points in my "en" and 3 in my "de" datasets.

The loss is now dropping appropriately. Currently at 7.207 for step 5,000.

Would probably explain my difficulties when trying to adapt https://github.com/Tomiinek/Multilingual_Text_to_Speech earlier. It would generate correct pronunciations, mostly, it had a tendency to run away at the end of sentences with mumbling noises.

**Comment by Flux9665** [comment]
The mumbling at the end of sentences is an inherent weakness for autoregressive models such as Tacotron, which the repo you linked is built on, maybe some careful hyperparameter tuning could even fix that. Errors propagate from start to end and that can cause also the stop-token to be predicted incorrectly. So yes, I also found that autoregressive models need pretty clean data to work.

But nice to hear that removing the flawed datapoints worked!

**Comment by Flux9665** [comment]
Scorer utility is added with today's release. It will report any NaNs it encounters and can show you the loss for each datapoint and its associated path. It might make sense to run this in a notebook and not in a script, since it iterates though all of the data and assigns each point a score and can then be queried afterwards, so an interactive environment could be nicer.

**Comment by kin0303** [comment]
> Had 5 bad data points in my "en" and 3 in my "de" datasets.

Hi, do you run the training in 2 languages?
